File: test/integration/python/test_add_prims.py
# -*- coding: utf-8 -*-

from pxr import Usd, Tf
from usd_notice_framework import Broker, UnfNotice

import pytest


@pytest.mark.parametrize("notice_type, excepted", [
    ("StageNotice", 6),
    ("StageContentsChanged", 3),
    ("ObjectsChanged", 3),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prims(notice_type, excepted):
    """Add several prims to the stage.
    """
    stage = Usd.Stage.CreateInMemory()
    Broker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(UnfNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    stage.DefinePrim("/Foo")
    stage.DefinePrim("/Bar")
    stage.DefinePrim("/Baz")

    # Ensure that we received the same number of notices.
    assert len(received_broker) == excepted
    assert len(received_usd) == excepted

@pytest.mark.parametrize("notice_type, expected_usd, expected_broker", [
    ("StageNotice", 6, 2),
    ("StageContentsChanged", 3, 1),
    ("ObjectsChanged", 3, 1),
    ("StageEditTargetChanged", 0, 0),
    ("LayerMutingChanged", 0, 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prims_batching(notice_type, expected_usd, expected_broker):
    """Add several prims to the stage and batch broker notices.
    """
    stage = Usd.Stage.CreateInMemory()
    broker = Broker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(UnfNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    broker.BeginTransaction()

    stage.DefinePrim("/Foo")
    stage.DefinePrim("/Bar")
    stage.DefinePrim("/Baz")

    # Ensure that broker notices are not sent during a transaction.
    assert len(received_broker) == 0

    # While USD Notices are being sent as expected.
    assert len(received_usd) == expected_usd

    broker.EndTransaction()

    # Ensure that consolidated broker notices are sent after a transaction.
    assert len(received_broker) == expected_broker

@pytest.mark.parametrize("notice_type, expected_usd", [
    ("StageNotice", 6),
    ("StageContentsChanged", 3),
    ("ObjectsChanged", 3),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prims_blocking(notice_type, expected_usd):
    """Add several prims to the stage and block broker notices.
    """
    stage = Usd.Stage.CreateInMemory()
    broker = Broker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(UnfNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Predicate blocking all broker notices.
    broker.BeginTransaction(predicate=lambda _: False)

    stage.DefinePrim("/Foo")
    stage.DefinePrim("/Bar")
    stage.DefinePrim("/Baz")

    # Ensure that broker notices are not sent during a transaction.
    assert len(received_broker) == 0

    # While USD Notices are being sent as expected.
    assert len(received_usd) == expected_usd

    broker.EndTransaction()

    # Ensure that no broker notices have been received.
    assert len(received_broker) == 0

def test_add_prims_transaction_objectschanged():
    """Add several prims to the stage during transaction and analyze
    ObjectsChanged notice.

    """
    stage = Usd.Stage.CreateInMemory()
    broker = Broker.Create(stage)

    received = []

    def _validate(notice, stage):
        """Validate notice received."""
        assert len(notice.GetResyncedPaths()) == 3
        assert len(notice.GetChangedInfoOnlyPaths()) == 0
        assert notice.GetResyncedPaths()[0] == "/Foo"
        assert notice.GetResyncedPaths()[1] == "/Bar"
        assert notice.GetResyncedPaths()[2] == "/Baz"
        received.append(notice)

    key = Tf.Notice.Register(UnfNotice.ObjectsChanged, _validate, stage)

    broker.BeginTransaction()

    stage.DefinePrim("/Foo")
    stage.DefinePrim("/Bar")
    stage.DefinePrim("/Baz")

    broker.EndTransaction()

    # Ensure that one notice was received.
    assert len(received) == 1

File: test/integration/testAddPrims.cpp
#include <unf/broker.h>

#include <unfTest/listener.h>
#include <unfTest/notice.h>
#include <unfTest/observer.h>

#include <gtest/gtest.h>
#include <pxr/usd/sdf/path.h>
#include <pxr/usd/usd/primRange.h>
#include <pxr/usd/usd/stage.h>

#include <typeinfo>

// namespace aliases for convenience.
using _USD = PXR_NS::UsdNotice;
namespace _UNF = unf::UnfNotice;

class AddPrimsTest : public ::testing::Test {
  protected:
    using UsdListener = ::Test::Listener<
        _USD::StageNotice, _USD::StageContentsChanged, _USD::ObjectsChanged,
        _USD::StageEditTargetChanged, _USD::LayerMutingChanged>;

    using BrokerListener = ::Test::Listener<
        _UNF::StageNotice, _UNF::StageContentsChanged,
        _UNF::ObjectsChanged, _UNF::StageEditTargetChanged,
        _UNF::LayerMutingChanged>;

    void SetUp() override
    {
        _stage = PXR_NS::UsdStage::CreateInMemory();
        _usdListener.SetStage(_stage);
        _brokerListener.SetStage(_stage);
    }

    PXR_NS::UsdStageRefPtr _stage;

    UsdListener _usdListener;
    BrokerListener _brokerListener;
};

TEST_F(AddPrimsTest, Simple)
{
    auto broker = unf::Broker::Create(_stage);

    _stage->DefinePrim(PXR_NS::SdfPath{"/Foo"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Bar"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Baz"});

    // Ensure that similar notices are received via the stage and the broker.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 6);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 0);

    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 6);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 3);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 3);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);
}

TEST_F(AddPrimsTest, Batching)
{
    auto broker = unf::Broker::Create(_stage);

    broker->BeginTransaction();

    _stage->DefinePrim(PXR_NS::SdfPath{"/Foo"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Bar"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Baz"});

    // Ensure that USD Notices are being sent as expected.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 6);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 0);

    // While broker notices are blocked during a transaction.
    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    // Ensure that consolidated broker notices are sent after a transaction.
    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 2);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 1);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 1);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);
}

TEST_F(AddPrimsTest, Blocking)
{
    auto broker = unf::Broker::Create(_stage);

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction([](const _UNF::StageNotice&) { return false; });

    _stage->DefinePrim(PXR_NS::SdfPath{"/Foo"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Bar"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Baz"});

    // Ensure that USD Notices are being sent as expected.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 6);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 0);

    // While broker notices are blocked during a transaction.
    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    // Ensure that no broker notices are sent after a transaction either.
    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);
}

TEST_F(AddPrimsTest, PartialBlocking)
{
    auto broker = unf::Broker::Create(_stage);

    std::string target = typeid(_UNF::ObjectsChanged).name();

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction([&](const _UNF::StageNotice& n) {
        return (n.GetTypeId() == target);
    });

    _stage->DefinePrim(PXR_NS::SdfPath{"/Foo"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Bar"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Baz"});

    // Ensure that USD Notices are being sent as expected.
    ASSERT_EQ(_usdListener.Received<_USD::StageNotice>(), 6);
    ASSERT_EQ(_usdListener.Received<_USD::StageContentsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::ObjectsChanged>(), 3);
    ASSERT_EQ(_usdListener.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_usdListener.Received<_USD::LayerMutingChanged>(), 0);

    // While broker notices are blocked during a transaction.
    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    // Ensure that only consolidated ObjectsChanged broker notice are sent.
    ASSERT_EQ(_brokerListener.Received<_UNF::StageNotice>(), 1);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageContentsChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::ObjectsChanged>(), 1);
    ASSERT_EQ(_brokerListener.Received<_UNF::StageEditTargetChanged>(), 0);
    ASSERT_EQ(_brokerListener.Received<_UNF::LayerMutingChanged>(), 0);
}

TEST_F(AddPrimsTest, Transaction_ObjectsChanged)
{
    auto broker = unf::Broker::Create(_stage);

    ::Test::Observer<_UNF::ObjectsChanged> observer(_stage);

    broker->BeginTransaction();

    _stage->DefinePrim(PXR_NS::SdfPath{"/Foo"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Bar"});
    _stage->DefinePrim(PXR_NS::SdfPath{"/Baz"});

    ASSERT_EQ(observer.Received(), 0);

    broker->EndTransaction();

    ASSERT_EQ(observer.Received(), 1);

    const auto& n = observer.GetLatestNotice();
    ASSERT_EQ(n.GetResyncedPaths().size(), 3);
    ASSERT_EQ(n.GetResyncedPaths().at(0), PXR_NS::SdfPath{"/Bar"});
    ASSERT_EQ(n.GetResyncedPaths().at(1), PXR_NS::SdfPath{"/Baz"});
    ASSERT_EQ(n.GetResyncedPaths().at(2), PXR_NS::SdfPath{"/Foo"});
    ASSERT_EQ(n.GetChangedInfoOnlyPaths().size(), 0);
}